llvm-project/Dockerfile

```diff
@@ -0,0 +1,133 @@
+# _Unwind_Resume issues with llvm-libunwind? Cmake can't find it either?
+# TODO: can we build stage1 with llvm binutils? llvm13 alpine package contains.
+#RUN apk add --no-cache clang lld compiler-rt musl-dev
+RUN apk add --no-cache clang lld musl-dev
+
+COPY --from=source llvm-project-14.0.1.src.tar.xz /
+RUN tar xf llvm-project-14.0.1.src.tar.xz && \
+  mv llvm-project-14.0.1.src llvm-project && \
+  mkdir llvm-project/llvm/build
+
+# libcxxabi depends on zlib-dev!
+# compiler-rt depends on libexecinfo-dev! (unless gwp-asan is disabled)
+# TODO: libxml2?
+# llvm cmake throws lots of warnings without git
+RUN apk add --no-cache cmake ninja python3 zlib-dev git
+
+# This seems like a mistake that clang++ in alpine cannot compile C++ hello
+# world.
+# TODO: this also pulls in gnu binutils...
+RUN apk add --no-cache g++ && \
+  find /usr/bin/ | grep -e gcc -e g++ | grep -v clang++ | xargs rm && \
+  rm /usr/bin/cpp /usr/bin/c99 /usr/bin/c89 && \
+  ln -sf /usr/bin/clang /usr/bin/cc
+
+# Test stage 0 clang and clang++.
+COPY hello.c hello.cpp /
+
+RUN clang hello.c && ./a.out && \
+  clang++ hello.cpp && ./a.out
+
+COPY stage1.cmake llvm-project/.
+RUN cmake \
+  -B llvm-project/llvm/build \
+  -C llvm-project/stage1.cmake \
+  -S llvm-project/llvm \
+  -G Ninja
+
+# While these could be one single command, generally for development it's very
+# painful when one of many specified targets fail as then nothing gets cached
+# by docker. Break these into distinct commands to optimize for development
+# iteration speed.
+RUN ninja -C llvm-project/llvm/build install-builtins
+RUN ninja -C llvm-project/llvm/build install-compiler-rt
+RUN ninja -C llvm-project/llvm/build install-unwind
+RUN ninja -C llvm-project/llvm/build install-cxxabi
+# Necessary for libcxx
+RUN apk add --no-cache linux-headers
+RUN ninja -C llvm-project/llvm/build install-cxx
+RUN ninja -C llvm-project/llvm/build install-clang install-lld
+RUN ninja -C llvm-project/llvm/build install-clang-resource-headers
+
+### END STAGE1
+### START STAGE2
+
+# STAGE2 goal, build WITHOUT apk add
+FROM alpine:edge as stage_two
+COPY --from=stage_one llvm-project/llvm/build/bin /usr/local/bin
+COPY --from=stage_one /usr/local/lib /usr/local/lib
+COPY --from=stage_one /usr/local/include /usr/local/include
+COPY hello.c hello.cpp /
+
+RUN cd /usr/lib/ && \
+  for library in libc++abi.so.1 libc++.a libc++abi.a libc++.so.1 libunwind.so.1 libunwind.a; \
+    do ln -s "/usr/local/lib/x86_64-alpine-linux-musl/${library}" . ; \
+  done
+
+# The stage 1 build of clang/lld/other binaries still depends on libstdc++.
+RUN apk add --no-cache libstdc++ musl-dev
+
+# Pause for a quick sanity check
+# Stage 1 clang only defaulted-on compiler-rt and lld
+RUN clang++ hello.cpp \
+  --stdlib=libc++ \
+  --rtlib=compiler-rt \
+  --unwindlib=libunwind \
+  -fuse-ld=lld && \
+  ./a.out
+
+RUN clang++ hello.cpp \
+  --stdlib=libc++ \
+  --unwindlib=libunwind && \
+  ./a.out
```

Comment:
The first test specifies the 4 flags we need to use the LLVM variants.

The second test verifies that 2 of the flags already have the implicit defaults we desire.  The second test does cover everything that the first does, so the first could be dropped. But I find the first test descriptive of what the 4 flags are, since I'll probably forget them in the near future.

Reply:
Sounds good, just wanted to make sure there wasn't something missing!